### PR TITLE
TensorGraph uses a persistent Session

### LINKED ISSUE #768: Save/restore for TensorGraph
In the past, our models have generally had a persistent Session object containing all the variables for the model.  They automatically wrote out checkpoints during training, and provided a `restore()` method to load the variables from the most recent checkpoint into memory.

TensorGraph works differently.  It creates a new Session every time you call `fit()` or `predict()`, loads the variables from the latest checkpoint, then throws away the Session before returning.  I find this behavior has a couple of problems.

First, it's really slow.  If you want to run prediction thousands of times, that means it will load the variables from disk thousands of times.  This is why for both RL and MAML I've basically had to subvert the design of TensorGraph.  I use it to build the Tensorflow graph, but when I need to do any calculations I ignore the corresponding parts of TensorGraph, pull out the needed internal fields, and do the calculations directly.

It also is inflexible, since it enforces that prediction can only ever be done based on the variables in the latest checkpoint file, no others.  A good example of where this causes problems is MAML.  It uses TensorGraph to define the model, but the actual optimization is done with a different optimizer and a different loss function.  That produces a model (which gets saved to disk) that is designed to be easy to train, but isn't optimized for any particular task.  To use it for prediction, you first do a few steps of gradient descent to produce a fine tuned version of the model (which does *not* get saved to disk), and do the prediction based on that.  But I can't use any of TensorGraph's prediction methods for it, because it would just throw away the tuned variables and replace them with the generic ones from disk.

I think we should consider changing this behavior.

### PR DESCRIPTION
**Pull request by peastman**
Fixes #768.

### CODE REVIEW

**Comment by peastman** [comment]
I think this is ready to merge, if no one sees any problems with it.

**Comment by lilleswing** [comment]
LGTM

Do we want to expose close() as a method on TensorGraph?  Or should we have users go tg.close().
Do we expect other long running cleanup that needs to be done?

**Comment by peastman** [comment]
If the TensorGraph gets garbage collected, all of its resources (the Graph and the Session) should automatically get cleaned up.  Are there situations where you would keep the TensorGraph around but still need those to be cleaned up?

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 1]
```diff
@@ -605,24 +605,21 @@ def predict_proba_on_generator(self, generator, transformers=[]):
...
+      out_tensors = [x.out_tensor for x in self.outputs]
+      results = []
+      for feed_dict in generator:
+        feed_dict = {
+            self.layers[k.name].out_tensor: v
+            for k, v in six.iteritems(feed_dict)
+        }
+        feed_dict[self._training_placeholder] = 1.0  ##
+        result = np.array(self.session.run(out_tensors, feed_dict=feed_dict))
+        if len(result.shape) == 3:
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Should we take the chance to remove the transpose here?

File: deepchem/models/tensorgraph/models/graph_models.py [hunk 2]
```diff
@@ -868,23 +865,20 @@ def predict_proba_on_generator(self, generator, transformers=[]):
...
+      for feed_dict in generator:
+        # Extract number of unique samples in the batch from w_b
+        n_valid_samples = len(np.nonzero(feed_dict[self.weights][:, 0])[0])
+        feed_dict = {
+            self.layers[k.name].out_tensor: v
+            for k, v in six.iteritems(feed_dict)
+        }
+        feed_dict[self._training_placeholder] = 0.0
+        result = np.array(self.session.run(out_tensors, feed_dict=feed_dict))
+        if len(result.shape) == 3:
```
Comment (rbharath) [inline deepchem/models/tensorgraph/models/graph_models.py]: Same comment about the transpose here.

**Comment by lilleswing** [comment]
@peastman I'm not sure it is garbage collected.  
https://www.tensorflow.org/api_docs/python/tf/Session

The second thing they tell you is how important it is to close the Session.

File: deepchem/models/tensorgraph/models/graph_models.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/tensorgraph/models/graph_models.py]: That's unrelated to this PR.  The only thing that changed about those lines was the indentation.

**Comment by peastman** [comment]
Session automatically calls `close()` when it's garbage collected: https://github.com/tensorflow/tensorflow/blob/r0.8/tensorflow/python/client/session.py#L139-L140

**Comment by lilleswing** [comment]
@peastman fair enough

**Comment by rbharath** [comment]
Cool, sounds like we're good to go.

LGTM
